File: ResumeFormatter/json-objects/resume.rb
require_relative 'contact'
require_relative 'skills'
require_relative 'project'
require_relative 'job'
require_relative 'education'

class Resume
  attr_accessor :name, :contact, :skills, :projects, :jobs, :education

  def initialize(hash)
    return unless hash.respond_to? :fetch

    @name      = hash.fetch('name',      nil)
    @contact   = hash.fetch('contact',   nil)
    @skills    = hash.fetch('skills',    nil)
    @projects  = hash.fetch('projects',  nil)
    @jobs      = hash.fetch('jobs',      nil)
    @education = hash.fetch('education', nil)

    @contact   = Contact.new @contact unless @contact.nil?
    @skills    = Skills.new  @skills  unless @skills.nil?
    @projects  = Entry.convert @projects, Project
    @jobs      = Entry.convert @jobs, Job
    @education = Entry.convert @education, Education
  end

end

File: formatter/ResumeFormatter/src/mpeschke/resume/Resume.scala
package mpeschke.resume

import io.circe.syntax._
import io.circe.{Decoder, Encoder, Json}
import mpeschke.resume.Resume.Name

final case class Resume(
                       name: Name,
                       contact: ContactInfo,
                       skills: Skills,
                       projects: List[Project],
                       jobs: List[Job],
                       education: List[Education]
                       )
object Resume {
  object Name extends StringWrapper
  type Name = Name.Type

  implicit final val encoder: Encoder[Resume] = Encoder.instance[Resume] { r =>
    Json.obj(
      "name" -> r.name.asJson,
      "contact" -> r.contact.asJson,
      "skills" -> r.skills.asJson,
      "projects" -> r.projects.asJson,
      "jobs" -> r.jobs.asJson,
      "education" -> r.education.asJson
    )
  }
  implicit final val decoder: Decoder[Resume] = Decoder.instance[Resume] { cursor =>
    for {
      name <- cursor.downField("name").as[Name]
      contact <- cursor.downField("contact").as[ContactInfo]
      skills <- cursor.downField("skills").as[Skills]
      projects <- cursor.downField("projects").as[List[Project]]
      jobs <- cursor.downField("jobs").as[List[Job]]
      education <- cursor.downField("education").as[List[Education]]
    } yield Resume(name, contact, skills, projects, jobs, education)
  }
}